### PR TITLE
add SQLDatabaseChain (only pg + sqlite right now) (#144)

### PR DESCRIPTION
**Pull request by agola11**
* add Sqldatabase.ts file to be able to create SQLChain

* implement feedback

* add serializer

* add Sqlite support + fix a bug

### CODE REVIEW

File: langchain/src/chains/sql_database/sql_database_chain.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  // Number of results to return from the query
+  topK = 5;
+
+  inputKey = "query";
+
+  outputKey = "result";
+
+  returnIntermediateSteps = false;
+
+  // Whether or not to return the intermediate steps along with the final answer.
```
Comment (nfcampos) [inline langchain/src/chains/sql_database/sql_database_chain.ts:36]: Move up

**Review by nfcampos — APPROVED** [review]
lgtm

File: langchain/src/chains/tests/sql_database_chain.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    `);
+  await datasource.query(`
+        INSERT INTO users (name, age) VALUES ('Alice', 20);
+    `);
+  await datasource.query(`
+        INSERT INTO users (name, age) VALUES ('Bob', 21);
+    `);
+  await datasource.query(`
+        INSERT INTO users (name, age) VALUES ('Charlie', 22);
+    `);
```
Comment (agola11) [inline langchain/src/chains/tests/sql_database_chain.int.test.ts:26]: @nfcampos -- FYI couldn't figure out a way to add a bunch of statements to the same `query` call (it only seemed to process the first one). not sure if there's a better way

File: langchain/src/sql_database.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    if (fields?.includesTables && fields?.ignoreTables) {
+      throw new Error("Cannot specify both include_tables and ignoreTables");
+    }
+    this.includesTables = fields?.includesTables ?? [];
+    this.ignoreTables = fields?.ignoreTables ?? [];
+    this.sampleRowsInTableInfo =
+      fields?.sampleRowsInTableInfo ?? this.sampleRowsInTableInfo;
+  }
+
+  static async fromDataSourceParams(
```
Comment (agola11) [inline langchain/src/sql_database.ts:41]: this was added to allow passing in `typeorm.DataSource` directly (made testing easier and some users might want that)

File: langchain/src/sql_database.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+  allTables: Array<SqlTable> = [];
+
+  includesTables: Array<string> = [];
+
+  ignoreTables: Array<string> = [];
+
+  sampleRowsInTableInfo = 3;
+
+  protected constructor(fields: SqlDatabaseDataSourceParams) {
```
Comment (agola11) [inline langchain/src/sql_database.ts:29]: made constructor protected bc we never want users instantiating this class with it (it wouldn't initialize the `DataSource` or `validateTables`). Those need to happen outside in a `from` static method bc they're async

File: langchain/src/chains/tests/sql_database_chain.int.test.ts (hunk 2 above)
Comment (nfcampos) [reply langchain/src/chains/tests/sql_database_chain.int.test.ts:26]: As far as I can tell there is no way to do that in typeorm...

File: docs/docs/modules/chains/sql_db.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    appDataSource: datasource,
+  });
+
+  const chain = new SqlDatabaseChain({
+    llm: new OpenAI({ temperature: 0 }),
+    database: db,
+  });
+
+  const res = await chain.run("How many tracks are there?");
+  console.log(res);
```
Comment (nfcampos) [inline docs/docs/modules/chains/sql_db.md:29]: Let's also add the db destroy to any examples @agola11

Comment (agola11) [reply docs/docs/modules/chains/sql_db.md:29]: will do! And will also add to tests

**Comment by nfcampos** [comment]
LGTM

**Comment by agola11** [comment]
Thank you for the great contribution @MaximeThoonsen 🙏🏽

**Pull request merged**